### YouTokenToMe tokenizer
https://github.com/vkcom/YouTokenToMe

In [1]:
!pip install youtokentome

     |████████████████████████████████| 1.7 MB 345 kB/s eta 0:00:01
     |████████████████████████████████| 82 kB 13.7 MB/s eta 0:00:01


In [2]:
BRYAK!!!!

SyntaxError: invalid syntax (<ipython-input-2-ba15253e5fe6>, line 1)

In [3]:
import os
import random
import glob
import numpy as np
from collections import deque
import subprocess
import youtokentome as yttm
import multiprocessing
import time

In [4]:
data_filepath = "/mnt/data5/docker/mini/share/spandh.ami1/data/rnnlm_data/1M.low.shuffled"

print("%s is file: %s" % (data_filepath, os.path.isfile(data_filepath)))

if os.path.isfile(data_filepath):
    print("File size: %.2f Mbytes" % (os.stat(data_filepath).st_size / 1024. / 1024.))

/mnt/data5/docker/mini/share/spandh.ami1/data/rnnlm_data/1M.low.shuffled is file: True
File size: 120.30 Mbytes


In [5]:
test_data_filepath = "/mnt/data5/docker/mini/share/spandh.ami1/data/rnnlm_data/test"

print("%s is file: %s" % (test_data_filepath, os.path.isfile(test_data_filepath)))

if os.path.isfile(test_data_filepath):
    print("File size: %.2f Mbytes" % (os.stat(test_data_filepath).st_size / 1024. / 1024.))

/mnt/data5/docker/mini/share/spandh.ami1/data/rnnlm_data/test is file: True
File size: 2845.08 Mbytes


In [6]:
VOCAB_SIZE = 4000
model_path = "example.model"

### Trains BPE model and saves to file.  


<code>youtokentome.BPE.train(data, model, vocab_size, coverage, n_threads=-1, pad_id=0, unk_id=1, bos_id=2, eos_id=3)</code>

Args:  
  
**data**: string, path to file with training data  
**model**: string, path to where the trained model will be saved  
**vocab_size**: int, number of tokens in the final vocabulary  
**coverage**: float, fraction of characters covered by the model. Must be in the range [0, 1]. A good value to use is about 0.9999.  
**n_threads**: int, number of parallel threads used to run. If -1 is passed, then all available threads are going to be used. Note that the number of threads is limited by 8 (see benchmark).  
**pad_id**: int, reserved id for padding  
**unk_id**: int, reserved id for unknown symbols  
**bos_id**: int, reserved id for begin of sentence token  
**eos_id**: int, reserved id for end of sentence token  
**Returns**: Class youtokentome.BPE with the loaded model.  

In [6]:
%%time
# Training model
yttm.BPE.train(data=data_filepath, vocab_size=VOCAB_SIZE, model=model_path)

CPU times: user 9.42 s, sys: 756 ms, total: 10.2 s
Wall time: 2.28 s


In [7]:
%%time
# Training model
yttm.BPE.train(data=test_data_filepath, vocab_size=VOCAB_SIZE, model=model_path)

CPU times: user 2min 22s, sys: 12.2 s, total: 2min 34s
Wall time: 30.5 s


### Model loading  


<code>youtokentome.BPE(model, n_threads=-1)</code> 

Class constructor. Loads the trained model.  
  
**model**: string, path to the trained model  
**n_threads**: int, number of parallel threads used to run. If equal to -1, then the maximum number of threads available will be used.

In [7]:
%%time
# Loading model
bpe = yttm.BPE(model=model_path)

CPU times: user 8 ms, sys: 4 ms, total: 12 ms
Wall time: 11.9 ms


In [8]:
test_file = open(test_data_filepath, "rt")
test_texts = [test_file.readline().strip() for i in range(10)]
test_file.close()

### Two types of tokenization

In [9]:
# 1
print(*bpe.encode(test_texts, output_type=yttm.OutputType.ID), sep="\n")

[107, 505, 3797, 1008, 1616, 3683, 129, 53, 884, 80, 3902, 276, 1915, 237, 362, 935, 104, 1615, 85, 912, 3021, 2350, 240, 193, 508, 57, 1949, 337, 80, 57, 973, 124, 57, 3013, 2571, 145, 2247, 107, 2939, 185, 1679, 85, 2133, 261, 104, 140, 2594, 86, 242, 299, 179, 57, 1739, 80, 53, 3013, 2877]
[256, 219, 2282, 68, 17, 180, 1341, 2648, 2715, 1124, 686, 11, 2602, 1933, 3001, 277, 2013, 1049, 476, 85, 335]
[250, 306, 250, 773, 501, 250, 773, 393, 57, 688, 155, 797, 1557, 85, 250, 375, 14, 538, 76, 553, 605, 10, 123, 124, 250, 589, 816, 219, 1882, 124, 392, 553, 1492, 103, 976, 267, 332, 1600, 362, 76, 1068, 85, 250, 294, 140, 3392, 2015]
[397, 335, 421, 196, 3925, 438, 106, 1885, 72, 249, 3050, 1903, 2211, 2368, 249, 95, 3854, 79, 1131, 213, 85, 57, 973, 124, 53, 1780, 3925, 155, 2416, 72, 86, 140, 3161, 60, 135, 241, 207, 106, 1010, 368, 1938, 11, 107, 53, 1885, 76, 57, 1029, 2211, 2368]
[57, 1419, 971, 11, 124, 740, 316, 17, 228, 89, 5, 147, 272, 7, 76, 699, 2999, 294, 2727, 86, 1168, 32

In [10]:
# 2
print(*bpe.encode(test_texts, output_type=yttm.OutputType.SUBWORD), sep="\n")

['▁for', '▁cent', 'uries', '▁american', '▁democ', 'racy', '▁as', '▁a', '▁process', '▁of', '▁conflict', '▁res', 'olution', '▁has', '▁been', '▁based', '▁on', '▁give', '▁and', '▁take', '▁negot', 'iation', '▁comp', 'rom', 'ise', '▁the', '▁accept', 'ance', '▁of', '▁the', '▁fact', '▁that', '▁the', '▁majority', '▁rules', '▁with', '▁respect', '▁for', '▁minor', 'ity', '▁rights', '▁and', '▁above', '▁all', '▁on', '▁an', '▁agreement', '▁to', '▁ab', 'ide', '▁by', '▁the', '▁results', '▁of', '▁a', '▁majority', '▁vote']
['▁they', '▁have', '▁created', '▁p', 'u', 'pp', 'ets', '▁veh', 'icles', '▁bas', 'ket', 's', '▁rob', 'ots', '▁jew', 'ell', 'ery', '▁build', 'ings', '▁and', '▁more']
['▁you', '▁said', '▁you', '▁did', '▁what', '▁you', '▁did', '▁when', '▁the', '▁group', '▁was', '▁very', '▁young', '▁and', '▁you', '▁imp', 'l', 'ied', '▁in', '▁your', '▁mem', 'o', 'ir', '▁that', '▁you', '▁could', "n't", '▁have', '▁done', '▁that', '▁if', '▁your', '▁super', 'st', 'ruct', 'ure', '▁had', '▁already', '▁been', '▁in'

In [11]:
# bpe.vocab()

In [12]:
# bpe.encode('i would like to know if darth vader was right', output_type=yttm.OutputType.SUBWORD)

### Создание TFRecords из очищенных текстовых корпусов, используя натренированный YouTokenToMe BPE токенайзер

In [13]:
import tensorflow as tf
tf.__version__

'2.0.1'

In [14]:
SEQ_LEN = 10
num_cores = 30

#### Многопоточная реализация

In [15]:
# Следующая функция может быть использована чтобы преобразовать значение в тип совместимый с
# с tf.Example.
def _intlist_to_int64_feature(value):
    """Преобразует bool / enum / int / uint в int64_list."""
    return tf.train.Feature(int64_list=tf.train.Int64List(value=value))

In [16]:
keys_to_features = {
    'X': tf.io.FixedLenFeature([SEQ_LEN], tf.int64),
    "y": tf.io.FixedLenFeature([1], tf.int64)
}

def _parse_function(example_proto):
  # Разберите `tf.Example` proto используя вышеприведенный словарь.
  return tf.io.parse_single_example(example_proto, keys_to_features)

def _parse_batch_function(example_proto):
  # Разберите `tf.Example` proto используя вышеприведенный словарь.
  return tf.io.parse_example(example_proto, keys_to_features)

In [17]:
# ОЧЕНЬ БЫСТРЫЙ ПОДСЧЕТ КОЛИЧЕСТВА СТРОК В ФАЙЛЕ
from itertools import (takewhile,repeat)

def rawincount(filename):
    f = open(filename, 'rb')
    bufgen = takewhile(lambda x: x, (f.raw.read(1024*1024) for _ in repeat(None)))
    return sum( buf.count(b'\n') for buf in bufgen )

In [18]:
%%time
rawincount(data_filepath)

CPU times: user 164 ms, sys: 28 ms, total: 192 ms
Wall time: 189 ms


1000000

In [19]:
def get_n_start_positions_of_textfile(textfilepath, num_pos):
    start_positions = []
    
    string_count = rawincount(textfilepath)
    rows_in_one_part = string_count // num_pos
#     print("Strings count:", string_count)
#     print("Rows in one part:", rows_in_one_part)

    opened_file = open(textfilepath, "rt")
    for i in range(num_pos-1):
        start_positions.append(opened_file.tell())
        if len(start_positions) == num_pos:
            break
        else:
            for j in range(rows_in_one_part):
                opened_file.readline()
    
    return start_positions, rows_in_one_part

In [20]:
start_positions, rows_in_one_part = get_n_start_positions_of_textfile(textfilepath=data_filepath, num_pos=num_cores)

In [21]:
rows_in_one_part

33333

In [22]:
start_positions

[0,
 4221058,
 8427350,
 12621918,
 16808412,
 20991602,
 25182434,
 29398134,
 33568594,
 37781198,
 41974097,
 46183838,
 50394787,
 54644410,
 58836947,
 63033386,
 67238410,
 71442983,
 75621227,
 79837691,
 84042568,
 88251386,
 92472181,
 96683396,
 100871607,
 105084280,
 109280218,
 113505763,
 117713854]

In [23]:
# TEST START POSITIONS
for pos in start_positions:
    opened_file = open(data_filepath, "rt")
    opened_file.seek(pos)
    print("\'" + opened_file.readline() + "\'\n")
    opened_file.close()

'the east cantons as a whole should therefore not be confused with the german language region created in nineteen sixty three or with the german speaking community of belgium which does not include the smaller malmedy and waimes areas
'

'we have also observed that many precalculus level students only employ mental action one and mental action two when asked to construct the graph of a dynamic function situation
'

'she had always loved playing classical violin which prompted her to leave the university after one year to attend a conservatory
'

'traffic clogged highways
'

'statement of right to rescind the sale length of trial period procedure for extending trial period procedure for rescinding sale and the date by which the hearing aid need to be returned to rescind the sale
'

'this institute is sponsored in part by the faculty development and instructional design centre and pepsi cola general bottlers incorporated
'

'the forty second amendment act of nineteen seventy six extended

In [24]:
def create_tfrecord_gziped(args):
    input_filepath, start_pos, lines_count, output_filepath = args
    
    opened_file = open(input_filepath, "rt")
    opened_file.seek(start_pos)
    
    # Create output directory if absent
    os.makedirs(os.path.dirname(output_filepath), exist_ok=True)
    
    options = tf.io.TFRecordOptions(compression_type="GZIP")
    writer = tf.io.TFRecordWriter(output_filepath + ".gzip", options=options)
    
    for _ in range(lines_count):
        line = opened_file.readline().strip()
        encoded_line = np.array(bpe.encode(line), dtype=np.uint16)
        if len(encoded_line) >= (SEQ_LEN + 1): # достаточно длинная строка
            for i in range(len(encoded_line)-SEQ_LEN):
                x, y = encoded_line[i:i+SEQ_LEN], encoded_line[i+SEQ_LEN: i+SEQ_LEN+1]
                feature = {
                    'X': _intlist_to_int64_feature(x),
                    'y': _intlist_to_int64_feature(y)
                }
                example = tf.train.Example(features=tf.train.Features(feature=feature))
                writer.write(example.SerializeToString())
    
    writer.close()
    opened_file.close()
    return True

In [25]:
def create_tfrecord(args):
    input_filepath, start_pos, lines_count, output_filepath = args
    
    opened_file = open(input_filepath, "rt")
    opened_file.seek(start_pos)
    
    # Create output directory if absent
    os.makedirs(os.path.dirname(output_filepath), exist_ok=True)
    
    writer = tf.io.TFRecordWriter(output_filepath)
    
    for _ in range(lines_count):
        line = opened_file.readline().strip()
        encoded_line = np.array(bpe.encode(line), dtype=np.uint16)
        if len(encoded_line) >= (SEQ_LEN + 1): # достаточно длинная строка
            for i in range(len(encoded_line)-SEQ_LEN):
                x, y = encoded_line[i:i+SEQ_LEN], encoded_line[i+SEQ_LEN: i+SEQ_LEN+1]
                feature = {
                    'X': _intlist_to_int64_feature(x),
                    'y': _intlist_to_int64_feature(y)
                }
                example = tf.train.Example(features=tf.train.Features(feature=feature))
                writer.write(example.SerializeToString())
    
    writer.close()
    opened_file.close()
    return True

In [26]:
%%time

datapath = test_data_filepath
start_positions, rows_in_one_part = get_n_start_positions_of_textfile(textfilepath=datapath, num_pos=num_cores)
output_directory = "data_gziped_test"

generator = (
    (datapath, start_pos, rows_in_one_part, "%s/tfrecord_%d_%d" % (output_directory, idx, start_pos)) 
    for idx, start_pos in enumerate(start_positions)
)

# ### DEBUG
# for i in generator:
#     print(i)

pool = multiprocessing.Pool(num_cores)
pool.map(create_tfrecord_gziped, generator)

sum_size = 0
for path in glob.glob("%s/*" % output_directory, recursive=True):
    sum_size += os.stat(path).st_size
print(sum_size / 1024. / 1024., "Megabytes")

('/mnt/data5/docker/mini/share/spandh.ami1/data/rnnlm_data/test', 0, 788573, 'data_gziped_test/tfrecord_0_0')
('/mnt/data5/docker/mini/share/spandh.ami1/data/rnnlm_data/test', 99475431, 788573, 'data_gziped_test/tfrecord_1_99475431')
('/mnt/data5/docker/mini/share/spandh.ami1/data/rnnlm_data/test', 198923787, 788573, 'data_gziped_test/tfrecord_2_198923787')
('/mnt/data5/docker/mini/share/spandh.ami1/data/rnnlm_data/test', 298424485, 788573, 'data_gziped_test/tfrecord_3_298424485')
('/mnt/data5/docker/mini/share/spandh.ami1/data/rnnlm_data/test', 397758274, 788573, 'data_gziped_test/tfrecord_4_397758274')
('/mnt/data5/docker/mini/share/spandh.ami1/data/rnnlm_data/test', 497190337, 788573, 'data_gziped_test/tfrecord_5_497190337')
('/mnt/data5/docker/mini/share/spandh.ami1/data/rnnlm_data/test', 596532745, 788573, 'data_gziped_test/tfrecord_6_596532745')
('/mnt/data5/docker/mini/share/spandh.ami1/data/rnnlm_data/test', 696051475, 788573, 'data_gziped_test/tfrecord_7_696051475')
('/mnt/dat

In [27]:
# %%time

# datapath = test_data_filepath
# start_positions, rows_in_one_part = get_n_start_positions_of_textfile(textfilepath=datapath, num_pos=num_cores)
# output_directory = "data_nonziped_test"

# generator = (
#     (datapath, start_pos, rows_in_one_part, "%s/tfrecord_%d_%d" % (output_directory, idx, start_pos)) 
#     for idx, start_pos in enumerate(start_positions)
# )

# pool = multiprocessing.Pool(num_cores)
# pool.map(create_tfrecord, generator)

# sum_size = 0
# for path in glob.glob("%s/*" % output_directory, recursive=True):
#     sum_size += os.stat(path).st_size
# print(sum_size / 1024. / 1024., "Megabytes")

In [28]:
# для входного 2845 мегабайтного файла
26573.187690734863 / 6134.21284198761

4.331963753987481

In [29]:
# для входной 120 мегабайтного файла
1124.935212135315 / 259.5092535018921

4.334855874906646

### По времени, компрессированный сделался даже чуток быстрее 33м42с против 34м26с
### По занимаемому месту, компрессированный весит в 4.33 раза меньше (6134.2 мегабайта против 26573.19 мегабайта)
### По отношению к оригиналу компрессированный занимает в ~2.16 раза больше, некомпрес в 9.34 раза больше

# ПРОВЕРИТЬ ПРАВИЛЬНО ЛИ СОЗДАЛИСЬ ФАЙЛЫ ДЛЯ ПОСЛЕДОВАТЕЛЬНОСТИ 10. ПОЧЕМУ ВЕС ПОЧТИ ТАКОЙЖЕ, ДАЖЕ ЧУТЬ МЕНЬШЕ !!!!!!!!!!!!!!!!!!!!!!!!


### Загрузка данных

In [30]:
%%time
# Компрессированные
input_directory = "/mnt/data5/bocharick/tf2.0rc0_tst_prj/BPE_token_tests/data_gziped_test2"

raw_dataset = tf.data.TFRecordDataset(
    filenames=glob.glob("%s/*" % input_directory, recursive=True), 
    compression_type="GZIP",
    num_parallel_reads=num_cores
).shuffle(buffer_size=200000)

# SLOW
# parsed_dataset = raw_dataset.map(_parse_function).batch(16000).prefetch(5)

# FAST
parsed_dataset = raw_dataset.batch(16000).map(_parse_batch_function).prefetch(5)

for i, jj in enumerate(parsed_dataset):
    pass
print(i, jj)

6972 {'X': <tf.Tensor: id=13991, shape=(13970, 10), dtype=int64, numpy=
array([[  68,   12,  101, ...,   76,  865,  913],
       [2598, 1315, 3786, ...,   54, 3508,  498],
       [ 206, 1301,  245, ...,  104, 1557,   83],
       ...,
       [ 493, 3583,  152, ..., 2366,  331,  690],
       [ 123,  112, 3420, ...,  417,   83, 1511],
       [  61,   27,  154, ...,  154,  585,  351]])>, 'y': <tf.Tensor: id=13992, shape=(13970, 1), dtype=int64, numpy=
array([[1445],
       [ 109],
       [ 932],
       ...,
       [ 193],
       [ 694],
       [1577]])>}
CPU times: user 22min 35s, sys: 12min 40s, total: 35min 15s
Wall time: 17min 46s


In [35]:
%%time
# Некомпрессированные
input_directory = "/mnt/data5/bocharick/tf2.0rc0_tst_prj/BPE_token_tests/data_nonziped_test2"

raw_dataset = tf.data.TFRecordDataset(
    filenames=glob.glob("%s/*" % input_directory, recursive=True), 
    num_parallel_reads=num_cores
).shuffle(buffer_size=200000)

# SLOW
# parsed_dataset = raw_dataset.map(_parse_function).batch(16000).prefetch(5)

# FAST
parsed_dataset = raw_dataset.batch(16000).map(_parse_batch_function).prefetch(5)

for i, jj in enumerate(parsed_dataset):
    pass
print(i, jj)

6972 {'X': <tf.Tensor: id=28333, shape=(13970, 10), dtype=int64, numpy=
array([[ 104,   50,  335, ..., 1309,  865, 1767],
       [  66, 1804,   82, ...,  126,  609,  126],
       [1891,   76,  923, ...,   80, 2757, 1812],
       ...,
       [1039, 2960, 1916, ..., 2710,  635,  203],
       [1187,  465, 1744, ..., 2303,  102,  663],
       [  54, 3744,  151, ...,  390,  819,   82]])>, 'y': <tf.Tensor: id=28334, shape=(13970, 1), dtype=int64, numpy=
array([[ 76],
       [363],
       [941],
       ...,
       [995],
       [278],
       [ 54]])>}
CPU times: user 22min 31s, sys: 12min 39s, total: 35min 10s
Wall time: 17min 14s


In [40]:
%%time
# Компрессированные
st_time = time.time()

input_directory = "/mnt/data5/bocharick/tf2.0rc0_tst_prj/BPE_token_tests/data_gziped_test2"

raw_dataset = tf.data.TFRecordDataset(
    filenames=glob.glob("%s/*" % input_directory, recursive=True), 
    compression_type="GZIP",
    num_parallel_reads=num_cores
).shuffle(buffer_size=200000)

# SLOW
# parsed_dataset = raw_dataset.map(_parse_function).batch(16000).prefetch(5)

# FAST
parsed_dataset = raw_dataset.batch(16000).map(_parse_batch_function).prefetch(5)

prev_time = time.time()
for i, jj in enumerate(parsed_dataset):
#     pass
    if i % 50 == 0:
        print("i = %d; time from start - %.3f seconds; time from prev - %.3f seconds" % (i, time.time() - st_time, time.time() - prev_time), flush=True)
        prev_time = time.time()
    if i == 1000:
        break
print(i, jj)

i = 0; time from start - 1.943 seconds; time from prev - 1.899 seconds
i = 50; time from start - 10.166 seconds; time from prev - 8.221 seconds
i = 100; time from start - 18.530 seconds; time from prev - 8.364 seconds
i = 150; time from start - 26.955 seconds; time from prev - 8.422 seconds
i = 200; time from start - 35.189 seconds; time from prev - 8.232 seconds
i = 250; time from start - 43.398 seconds; time from prev - 8.208 seconds
i = 300; time from start - 51.549 seconds; time from prev - 8.147 seconds
i = 350; time from start - 59.747 seconds; time from prev - 8.197 seconds
i = 400; time from start - 67.737 seconds; time from prev - 7.989 seconds
i = 450; time from start - 75.837 seconds; time from prev - 8.099 seconds
i = 500; time from start - 84.094 seconds; time from prev - 8.255 seconds
i = 550; time from start - 91.957 seconds; time from prev - 7.862 seconds
i = 600; time from start - 99.868 seconds; time from prev - 7.909 seconds
i = 650; time from start - 107.963 seconds

In [37]:
%%time
# Компрессированные
st_time = time.time()

input_directory = "/mnt/data5/bocharick/tf2.0rc0_tst_prj/BPE_token_tests/data_gziped_test"

raw_dataset = tf.data.TFRecordDataset(
    filenames=glob.glob("%s/*" % input_directory, recursive=True), 
    compression_type="GZIP",
    num_parallel_reads=num_cores
).shuffle(buffer_size=200000)

# SLOW
# parsed_dataset = raw_dataset.map(_parse_function).batch(16000).prefetch(5)

# FAST
parsed_dataset = raw_dataset.batch(16000).map(_parse_batch_function).prefetch(5)

prev_time = time.time()
for i, jj in enumerate(parsed_dataset):
#     pass
    if i % 50 == 0:
        print("i = %d; time from start - %.3f seconds; time from prev - %.3f seconds" % (i, time.time() - st_time, time.time() - prev_time), flush=True)
        prev_time = time.time()
    if i == 1000:
        break
print(i, jj)

i = 0; time from start - 2.462 seconds; time from prev - 2.421 seconds
i = 50; time from start - 11.647 seconds; time from prev - 9.183 seconds
i = 100; time from start - 19.937 seconds; time from prev - 8.287 seconds
i = 150; time from start - 30.652 seconds; time from prev - 10.713 seconds
i = 200; time from start - 40.803 seconds; time from prev - 10.148 seconds
i = 250; time from start - 50.007 seconds; time from prev - 9.202 seconds
i = 300; time from start - 58.699 seconds; time from prev - 8.688 seconds
i = 350; time from start - 67.399 seconds; time from prev - 8.698 seconds
i = 400; time from start - 76.400 seconds; time from prev - 8.999 seconds
i = 450; time from start - 85.277 seconds; time from prev - 8.876 seconds
i = 500; time from start - 94.432 seconds; time from prev - 9.153 seconds
i = 550; time from start - 103.219 seconds; time from prev - 8.783 seconds
i = 600; time from start - 112.284 seconds; time from prev - 9.061 seconds
i = 650; time from start - 121.608 sec

In [38]:
%%time
# Компрессированные
st_time = time.time()

input_directory = "/mnt/data5/bocharick/tf2.0rc0_tst_prj/BPE_token_tests/data_gziped_test"

raw_dataset = tf.data.TFRecordDataset(
    filenames=glob.glob("%s/*" % input_directory, recursive=True), 
    compression_type="GZIP",
    num_parallel_reads=num_cores
).shuffle(buffer_size=200000)

# SLOW
# parsed_dataset = raw_dataset.map(_parse_function).batch(16000).prefetch(5)

# FAST
parsed_dataset = raw_dataset.batch(16000).map(_parse_batch_function).prefetch(num_cores)

prev_time = time.time()
for i, jj in enumerate(parsed_dataset):
#     pass
    if i % 50 == 0:
        print("i = %d; time from start - %.3f seconds; time from prev - %.3f seconds" % (i, time.time() - st_time, time.time() - prev_time), flush=True)
        prev_time = time.time()
    if i == 1000:
        break
print(i, jj)

i = 0; time from start - 2.438 seconds; time from prev - 2.402 seconds
i = 50; time from start - 11.610 seconds; time from prev - 9.169 seconds
i = 100; time from start - 22.509 seconds; time from prev - 10.898 seconds
i = 150; time from start - 31.268 seconds; time from prev - 8.755 seconds
i = 200; time from start - 40.083 seconds; time from prev - 8.813 seconds
i = 250; time from start - 48.960 seconds; time from prev - 8.875 seconds
i = 300; time from start - 57.825 seconds; time from prev - 8.860 seconds
i = 350; time from start - 66.848 seconds; time from prev - 9.022 seconds
i = 400; time from start - 76.172 seconds; time from prev - 9.320 seconds
i = 450; time from start - 85.222 seconds; time from prev - 9.048 seconds
i = 500; time from start - 95.095 seconds; time from prev - 9.870 seconds
i = 550; time from start - 104.203 seconds; time from prev - 9.105 seconds
i = 600; time from start - 113.450 seconds; time from prev - 9.245 seconds
i = 650; time from start - 122.842 seco

In [39]:
%%time
# Компрессированные
st_time = time.time()

input_directory = "/mnt/data5/bocharick/tf2.0rc0_tst_prj/BPE_token_tests/data_gziped_test2"

raw_dataset = tf.data.TFRecordDataset(
    filenames=glob.glob("%s/*" % input_directory, recursive=True), 
    compression_type="GZIP",
    num_parallel_reads=num_cores
).shuffle(buffer_size=200000)

# SLOW
parsed_dataset = raw_dataset.map(_parse_function).batch(16000).prefetch(5)

# FAST
# parsed_dataset = raw_dataset.batch(16000).map(_parse_batch_function).prefetch(5)

prev_time = time.time()
for i, jj in enumerate(parsed_dataset):
#     pass
    if i % 50 == 0:
        print("i = %d; time from start - %.3f seconds; time from prev - %.3f seconds" % (i, time.time() - st_time, time.time() - prev_time), flush=True)
        prev_time = time.time()
    if i == 1000:
        break
print(i, jj)

i = 0; time from start - 2.859 seconds; time from prev - 2.729 seconds
i = 50; time from start - 60.463 seconds; time from prev - 57.603 seconds
i = 100; time from start - 109.824 seconds; time from prev - 49.358 seconds
i = 150; time from start - 160.080 seconds; time from prev - 50.250 seconds
i = 200; time from start - 208.137 seconds; time from prev - 48.056 seconds
i = 250; time from start - 257.311 seconds; time from prev - 49.173 seconds
i = 300; time from start - 312.425 seconds; time from prev - 55.112 seconds
i = 350; time from start - 359.356 seconds; time from prev - 46.930 seconds
i = 400; time from start - 406.856 seconds; time from prev - 47.498 seconds
i = 450; time from start - 454.884 seconds; time from prev - 48.027 seconds
i = 500; time from start - 505.873 seconds; time from prev - 50.987 seconds
i = 550; time from start - 554.900 seconds; time from prev - 49.025 seconds
i = 600; time from start - 607.513 seconds; time from prev - 52.611 seconds
i = 650; time from 

### Чтение файлов:  
Из датасета читалось 1000 батчей, каждый размером 16000 записей, с шафлингом датасета .shuffle(buffer_size=200000).  
В итоге обнаружил, что, оказывается, можно еще и "неправильно" читать данные в плане скорости.  
Конкретно в данном случае существенная разница в скорости была из за того что сначала в медленном, как выяснилось, варианте, бралась запись из файла, затем десериализовалась, и затем набирался батч. А если же взять недесериализованные данные, собрать из них батч, а потом уже весь батч десериализовать, получается значительная прибавка в производительности.  
  
Чтение медленным способом (парсинг каждой записи):  
#### 1) Чтение из 7 gzip файлов в num_parallel_reads=30, prefetch(5) за 16м47с

Чтение быстрым способом (парсинг батчей):  
#### 1) Чтение из 7 gzip файлов в num_parallel_reads=30, prefetch(5) за 2м44с  
#### 2) Чтение из 30 gzip файлов в num_parallel_reads=30, prefetch(5) за 3м7с  
#### 3) Чтение из 30 gzip файлов в num_parallel_reads=30, prefetch(30) за 3м6с  

То есть размер prefetch не влияет в данном случае на скорость чтения, а бОльшее количество файлов с параллельным чтением, видимо увеличило нагрузку на память, так как параллельно идет распаковка на лету всеми файлами. По идее имеет смысл брать компромиссное количество, а брать совсем максимальное во все ядра может и не требуется. Суть просто в том, что когда у тебя N параллельных файлов, то в одном батче набираются строки записи из каждого из этих файлов. То есть, если у тебя 30 файлов, а в батче 60 записей, то в итоге в батче будет по 2 строки из каждого из 30 файлов. Своего рода шаффлинг обеспечиваемый загрузкой, помимо еще шаффла который можно принудительно указать для tf.data.Dataset.  

Итого получается, что чтение в одной и той же ситуации, но с разницей в сериализации построчно или побатчево отличается в ~6.14 раз по скорости.  

Это все, конечно, по факту знания уже становится очевидно, но изначально, когда ковырялся в доках датасета, там десериализацию показывали именно построчно, и как-то не этот факт вообще внимания не обращал, до тех пор пока просто не показалось, что чтение достаточно медленное.  
